File: cdm-sources/src/cdm/base/staticdata/asset/common/functions/GetCashCurrency.py
# pylint: disable=line-too-long, invalid-name, missing-function-docstring, missing-module-docstring, superfluous-parens
# pylint: disable=wrong-import-position, unused-import, unused-wildcard-import, wildcard-import, wrong-import-order, missing-class-docstring
from __future__ import annotations
import sys
import datetime
import inspect
from decimal import Decimal
from rosetta.runtime.utils import *
from rosetta.runtime.func_proxy import replaceable, create_module_attr_guardian
from cdm.base.staticdata.asset.common.CurrencyCodeEnum import CurrencyCodeEnum
from cdm.base.staticdata.asset.common.AssetIdTypeEnum import AssetIdTypeEnum
from cdm.base.staticdata.asset.common.Cash import Cash
from cdm.base.staticdata.asset.common.functions.AssetIdentifierByType import AssetIdentifierByType

__all__ = ['GetCashCurrency']


@replaceable
def GetCashCurrency(cash: Cash) -> CurrencyCodeEnum:
    """
    Returns the enumerated Currency Code of a Cash asset.
    
    Parameters 
    ----------
    cash : Cash
    
    Returns
    -------
    currencyEnum : CurrencyCodeEnum
    
    """
    self = inspect.currentframe()
    
    
    cashId = AssetIdentifierByType(rosetta_resolve_attr(rosetta_resolve_attr(self, "cash"), "identifier"), rosetta_resolve_attr(AssetIdTypeEnum, "CURRENCY_CODE"))
    currencyEnum =  CurrencyCodeEnum(rosetta_resolve_attr(rosetta_resolve_attr(self, "cashId"), "identifier")[0])
    
    
    return currencyEnum

sys.modules[__name__].__class__ = create_module_attr_guardian(sys.modules[__name__].__class__)

File: cdm-sources/cdm/base/staticdata/asset/common/functions/GetCashCurrency.java
package cdm.base.staticdata.asset.common.functions;

import cdm.base.staticdata.asset.common.AssetIdTypeEnum;
import cdm.base.staticdata.asset.common.AssetIdentifier;
import cdm.base.staticdata.asset.common.Cash;
import cdm.base.staticdata.asset.common.CurrencyCodeEnum;
import com.google.inject.ImplementedBy;
import com.rosetta.model.lib.functions.RosettaFunction;
import com.rosetta.model.lib.mapper.MapperC;
import com.rosetta.model.lib.mapper.MapperS;
import com.rosetta.model.metafields.FieldWithMetaString;
import javax.inject.Inject;


@ImplementedBy(GetCashCurrency.GetCashCurrencyDefault.class)
public abstract class GetCashCurrency implements RosettaFunction {
	
	// RosettaFunction dependencies
	//
	@Inject protected AssetIdentifierByType assetIdentifierByType;

	/**
	* @param cash 
	* @return currencyEnum 
	*/
	public CurrencyCodeEnum evaluate(Cash cash) {
		CurrencyCodeEnum currencyEnum = doEvaluate(cash);
		
		return currencyEnum;
	}

	protected abstract CurrencyCodeEnum doEvaluate(Cash cash);

	protected abstract MapperC<? extends AssetIdentifier> cashId(Cash cash);

	public static class GetCashCurrencyDefault extends GetCashCurrency {
		@Override
		protected CurrencyCodeEnum doEvaluate(Cash cash) {
			CurrencyCodeEnum currencyEnum = null;
			return assignOutput(currencyEnum, cash);
		}
		
		protected CurrencyCodeEnum assignOutput(CurrencyCodeEnum currencyEnum, Cash cash) {
			currencyEnum = cashId(cash).<FieldWithMetaString>map("getIdentifier", assetIdentifier -> assetIdentifier.getIdentifier())
				.first().<String>map("Type coercion", fieldWithMetaString -> fieldWithMetaString == null ? null : fieldWithMetaString.getValue()).checkedMap("to-enum", CurrencyCodeEnum::fromDisplayName, IllegalArgumentException.class).get();
			
			return currencyEnum;
		}
		
		@Override
		protected MapperC<? extends AssetIdentifier> cashId(Cash cash) {
			return MapperC.<AssetIdentifier>of(assetIdentifierByType.evaluate(MapperS.of(cash).<AssetIdentifier>mapC("getIdentifier", _cash -> _cash.getIdentifier()).getMulti(), AssetIdTypeEnum.CURRENCY_CODE));
		}
	}
}
